skema/text_reading/text_reading/grounding/src/main/scala/org/ml4ai/grounding/MiraEmbeddingsGrounder.scala

```diff
@@ -52,14 +58,19 @@ class MiraEmbeddingsGrounder(groundingConcepts:Seq[GroundingConcept], embeddings
     val topKIndices = sortedIndices.take(k)
     val topSimilarities = sortedCosineSimilarities.take(k)
     val topConcepts = topKIndices.map(groundingConcepts)
+
     (topConcepts zip topSimilarities) map {
       case (concept, similarity) => GroundingCandidate(concept, similarity)
     }
   }
 
-  def getNormalizedDistance(text:String, name:String): Float = {
+  private val editDistancesCache = mutable.HashMap[(String, String), Float]()
+  private def getNormalizedDistance(text:String, name:String): Float = {
+
     def getDistance(source: String, target: String) = MED(source, target, allowSubstitute = false, allowTranspose = false, allowCapitalize = false).getDistance
-    return getDistance(text.toLowerCase(), name.toLowerCase())/text.length().max(name.length())
+
+    editDistancesCache.getOrElseUpdate((text, name), getDistance(text.toLowerCase(), name.toLowerCase())/text.length().max(name.length()))
```

Comment:
I wonder whether the `(text, name)` should be lowercased in the key, because it is that way in the value and otherwise duplicates might be stored.

Reply:
Indeed, also the order shouldn't be relevant for hashing the tuple. Do you know if there is a class that has a position invariant hash?

Reply:
I think they'd say that you have to store your own type and provide the hash function for it yourself.  Instead of that, you could just sort (lowercased) text and name.  I like the limited-size LRU.